test/metabase/lib/swap_test.cljc

```diff
@@ -113,3 +113,83 @@
         (is (=? [{:level :warn, :message #"Ambiguous match for clause in swap-clauses \[:sum .*"}]
                 (messages)))))))
+
+(deftest ^:parallel swap-clauses-aggregations-with-downstream-refs-test
+  (testing "swap-clauses updates downstream refs even if names change"
+    (let [base                 (-> (lib/query meta/metadata-provider (meta/table-metadata :orders))
+                                   (lib/aggregate (lib/sum (meta/field-metadata :orders :subtotal)))
+                                   (lib/aggregate (lib/sum (meta/field-metadata :orders :total)))
+                                   (lib/aggregate (lib/sum (meta/field-metadata :orders :tax)))
+                                   lib/append-stage)
+          ;; These three aggregations end up being named sum, sum_2 and sum_3 in later stages.
+          aggs-by-id           (->> (lib/aggregations base 0)
+                                    m/indexed
+                                    (into {}))
+          [subtotal total tax] (lib/visible-columns base)]
+      (doseq [[i1 agg1] aggs-by-id
+              [i2 agg2] aggs-by-id
+              :when (not= i1 i2)
+              :let [before (-> base
+                               (lib/filter (lib/< subtotal 100))
+                               (lib/filter (lib/< total    200))
+                               (lib/filter (lib/< tax      300)))
+                    after  (lib/swap-clauses before 0 agg1 agg2)]]
+        (testing "the aggregations really did get swapped"
+          (let [exp-aggs (merge aggs-by-id
+                                {i1 agg2
+                                 i2 agg1})]
+            (is (=? [(get exp-aggs 0) (get exp-aggs 1) (get exp-aggs 2)]
+                    (lib/aggregations after 0)))))
+        ;; The refs in the three filters should match the same columns as before, even though the names have changed
+        ;; when swapped.
+        (let [filter-cols (fn [query]
+                            (let [cols (lib/visible-columns query)]
+                              (for [[_< _opts agg _value] (lib/filters query -1)]
+                                (lib/find-matching-column agg cols))))
+              cols-before (filter-cols before)
+              cols-after  (filter-cols after)]
+          (is (every? some? cols-before))
+          (is (every? some? cols-after))
+          (is (every? :lib/source-uuid cols-before))
+          (is (every? :lib/source-uuid cols-after))
+          (testing "the ref always points at the correct underlying column"
+            (is (=? (map :lib/source-uuid cols-before)
+                    (map :lib/source-uuid cols-after)))
+            (testing "even though the names have changed where they were swapped"
+              (let [dcas-before (mapv :lib/desired-column-alias cols-before)
+                    dcas-after  (mapv :lib/desired-column-alias cols-after)
+                    i-unchanged (first (remove #{i1 i2} (range 3)))]
```

Comment:
I would prefer naming the magic `3` and maybe even not assuming that only one index is unchanged.

Reply:
Done.